src/ema/math.rs

```diff
@@ -0,0 +1,114 @@
+    volume_ema(prev, exp_complement, incoming, exp_smoothing)
+}
+
+/// Calculate the smoothing factor and its complement for a given combination of period and
+/// iterations.
+pub fn exp_smoothing_and_complement(smoothing: FixedU128, iterations: u32) -> (FixedU128, FixedU128) {
+    debug_assert!(smoothing <= Price::one());
+    let complement = Price::one() - smoothing;
+    // in order to determine the iterated smoothing factor we exponentiate the complement
+    let exp_complement = complement.saturating_pow(iterations as usize);
+    debug_assert!(exp_complement <= Price::one());
+    let exp_smoothing = Price::one() - exp_complement;
+    (exp_smoothing, exp_complement)
+}
+
+/// Calculates smoothing factor alpha for an exponential moving average based on `period`:
+/// `alpha = 2 / (period + 1)`. It leads to the "center of mass" of the EMA corresponding to be the
+/// "center of mass" of a `period`-length SMA.
+///
+/// Possible alternatives for `alpha = 2 / (period + 1)`:
+/// + `alpha = 1 - 0.5^(1 / period)` for a half-life of `period` or
+/// + `alpha = 1 - 0.5^(2 / period)` to have the same median as a `period`-length SMA. See
+/// https://en.wikipedia.org/wiki/Moving_average#Relationship_between_SMA_and_EMA (N = period).
+pub fn smoothing_from_period(period: u64) -> FixedU128 {
+    FixedU128::saturating_from_rational(2u64, period.max(1).saturating_add(1))
+}
+
+/// Calculate the next exponential moving average for the given prices.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+pub fn price_ema(prev: Price, prev_weight: FixedU128, incoming: Price, weight: FixedU128) -> Price {
+    debug_assert!(prev_weight + weight == Price::one());
+    // Safe to use bare `+` and `*` because `prev_weight + weight == 1`.
+    prev * prev_weight + incoming * weight
+}
+
+/// Calculate the next exponential moving average for the given balances.
+/// `prev` is the previous oracle value, `incoming` is the new value to integrate.
+/// `weight` is the weight of the new value, `prev_weight` is the weight of the previous value.
+///
+/// Note: Incurs rounding errors iff exactly one of the balance values is small and the other is
+/// big, but the relative error is small because the other value is greater than `u64::MAX`.
+pub fn balance_ema(prev: Balance, prev_weight: FixedU128, incoming: Balance, weight: FixedU128) -> Balance {
+    debug_assert!(prev_weight + weight == Price::one());
+    // Safe to use bare `+` and `*` because `prev_weight + weight == 1`.
+    // Equivalent to `prev_value * prev_weight + incoming_value * weight`
+    if prev < Balance::from(u64::MAX) && incoming < Balance::from(u64::MAX) {
+        // We use `*` in combination with `FixedU128::from` to avoid rounding errors induced by
+        // using `mul_int` with small values.
+        (prev_weight * FixedU128::from(prev) + weight * FixedU128::from(incoming)).saturating_mul_int(Balance::one())
```

Comment:
> It might be slightly better to do O_(t + k) = x_t + W^k * (O_t - x_t)

You're suggesting this only for small balance values, though, I assume?
Because the other branch below deals with values that might not fit in FixedI128.

Reply:
Added an invariant test using `rug` and didn't find any failure cases with `100_000` iterations.

Reply:
I am not seeing the situation in which O_(t + k) = (1 - W^k) * x_t + W^k * O_t is better than O_(t + k) = x_t + W^k * (O_t - x_t)

Are you referring to the requirement to do some logic to handle the subtraction with unsigned types (i.e. to technically handle O_t < x_t and O_t > x_t as separate cases)? I think this is better then allowing a higher rounding error by doing two multiplications, personally.

Reply:
So to be clear, what I am suggesting is to *not* branch based on whether the value fits in FixedI128. I suppose you would branch based on O_t < x_t or O_t > x_t. But the formula I'm proposing would be used either way.

Reply:
I think I actually mixed up the weights here, although I don't think it affects the point I'm making.

I think that in the current implementation, you are doing O_(t + k) = (1 - W^k) * O_t + W^k * x_t
And I would recommend O_(t + k) = O_t + W^k * (x_t - O_t)

Basically the way it's written here, W = alpha, whereas the way I wrote it in the initial comment, I was accidentally using W = 1 - alpha. Sorry for any confusion, let me know if this helps.

Reply:
Ok, will try it out.

Reply:
I like that it is automatically biased towards the historical oracle value in a balanced way :+1: